Update the ULFM Readme (#9927)
open-mpi/ompi: docs/features/ulfm.rst

This chapter documents the features and options specific to the **User
Level Failure Mitigation (ULFM)** Open MPI implementation.

> TL;DR
> -----

Reviewer: ```suggestion
Quick start
-----------
```

This is an extremely terse summary of how to use ULFM:

.. code-block::

>   ./configure --with-ft=ulfm [...options...]
>   make [-j N] all install
>   mpicc my-ft-program.c -o my-ft-program
>   mpiexec -n 4 --with-ft ulfm my-ft-program

Reviewer: ```suggestion
  $ ./configure --with-ft=ulfm [...options...]
  $ make [-j N] all install
  $ mpicc my-ft-program.c -o my-ft-program
  $ mpiexec -n 4 --with-ft ulfm my-ft-program
```

^^^^^^^^^^^^^^^^^^^^^^^^

You can launch your application with fault tolerance by simply using
the normal Open MPI ``mpirun`` launcher, with the
``--with-ft ulfm`` CLI option (or its synonym ``--with-ft mpi``):

.. code-block::

   shell$ mpirun --with-ft ulfm ...

>   .. important:: By default, fault tolerance is not active at run time.

Reviewer: ```suggestion
   .. important: By default, fault tolerance is not active at run time.
```

Author: This change doesn't look right, what is the defect it tries to fix?

Reviewer: I'm not sure what I was thinking here.  Let's disregard this comment...

>                  It must be enabled via `--with-ft ulfm`.

Reviewer: ```suggestion
                 It must be enabled via ``--with-ft ulfm``.
```

Open MPI level options
~~~~~~~~~~~~~~~~~~~~~~

Some default values are applied to some Open MPI parameters when using
``mpiexec --with-ft ulfm``. These defaults are obtained from the ``ft-mpi``
aggregate MCA param file
``$installdir/share/openmpi/amca-param-sets/ft-mpi``. You can tune the
runtime behavior with ULFM by either setting or unsetting variables in
this file, or by overiding the variable on the command line (e.g.,
``--mca btl ofi,self``).

  .. important:: Note that if fault tolerance is disabled at runtime,
                 that is, when not using ``--with-ft ulfm``), the
                 ``ft-mpi`` MCA param file is not loaded, thus
                 components that are unsafe for fault tolerance will
                 load normally (this may change observed performance
                 when comparing with and without fault tolerance).

* ``mpi_ft_enable <true|false> (default: false)``
  permits turning on/off fault tolerance at runtime. This option is
  automatically set to ``true`` from the aggregate MCA param file
  ``ft-mpi`` loaded when using ``--with-ft ulfm``. When false, failure

>   detection is disabled; Interfaces defined by the fault tolerance extensions

Author: I'd rather keep the ; as this is a list of statements for the `when false` case.

this file, or by overiding the variable on the command line (e.g.,
``--mca btl ofi,self``).

  .. important:: Note that if fault tolerance is disabled at runtime,
                 (that is, when not using ``--with-ft ulfm``), the
                 ``ft-mpi`` AMCA param file is not loaded, thus
                 components that are unsafe for fault tolerance will
                 load normally (this may change observed performance
                 when comparing with and without fault tolerance).

* ``mpi_ft_enable <true|false> (default: false)``
  permits turning on/off fault tolerance at runtime. This option is
  automatically set to ``true`` from the aggregate MCA param file
  ``ft-mpi`` loaded when using ``--with-ft ulfm``. When false, failure
  detection is disabled; Interfaces defined by the fault tolerance extensions
  are substituted with dummy non-fault tolerant implementations (e.g.,
  ``MPIX_Comm_agree`` is implemented with ``MPI_Allreduce``); All other
  controls below become irrelevant.
* ``mpi_ft_verbose <int> (default: 0)`` increases the output of the
  fault tolerance activities. A value of 1 will report detected
  failures.

> * ``mpi_ft_detector <true|false> (default: false)``, **DEPRECATED**

Reviewer: This "DEPRECATED" label might be better as a `.. caution:`, or somesuch.

  latency (typically 1us increase).* You may want to **enable this
  option if you experience false positive** processes incorrectly
  reported as failed with the Open MPI failure detector.

>   This option is only relevant when `mpi_ft_detector` is `true`.

Reviewer: ```suggestion
  This option is only relevant when ``mpi_ft_detector`` is ``true``.
```

* ``mpi_ft_detector_period <float> (default: 3e0 seconds)`` heartbeat
  period. Recommended value is 1/3 of the timeout. _Values lower than
  100us may impart a noticeable effect on latency (typically a 3us
  increase)._

>   This option is only relevant when `mpi_ft_detector` is `true`.

Reviewer: ```suggestion
  This option is only relevant when ``mpi_ft_detector`` is ``true``.
```

* ``mpi_ft_detector_timeout <float> (default: 1e1 seconds)`` heartbeat
  timeout (i.e. failure detection speed). Recommended value is 3 times
  the heartbeat period.

>   This option is only relevant when `mpi_ft_detector` is `true`.

Reviewer: ```suggestion
  This option is only relevant when `mpi_ft_detector` is ``true``.
```

Known Limitations in ULFM
-------------------------

* InfiniBand support is provided through the UCT BTL; fault tolerant
  operation over the UCX PML is not yet supported for production runs.
* TOPO, FILE, RMA are not fault tolerant. They are expected to work
  properly before the occurence of the first failure.

Modified, Untested and Disabled Components
------------------------------------------

> Frameworks and components which are not listed in the following list
> are unmodified and support fault tolerance. Listed frameworks may be
> **modified** (and work after a failure), **untested** (and work before
> a failure, but may malfunction after a failure), or **disabled** (they
> cause unspecified behavior all around when FT is enabled).

Reviewer: ```suggestion
Frameworks and components are listed below and categorized into one of three classifications:

1. **Modified:** This framework/component has been specifically modified such that it will continue to work after a failure.
2. **Untested:** This framework/component has not been modified and/or tested with fault tolerance scenarios, and _may_ malfunction after a failure.
3. **Disabled:** This framework/component will cause unspecified behavior when fault tolerance is enabled.

Any framework or component not listed below are categorized as **Unmodified**, meaning that it is unmodified for fault tolerance, but will continue to work correctly after a failure.```

> All runtime disabled components are listed in the ``ft-mpi`` aggregate
> MCA param file
> ``$installdir/share/openmpi/amca-param-sets/ft-mpi``. You can tune the
> runtime behavior with ULFM by either setting or unsetting variables in
> this file (or by overiding the variable on the command line (e.g.,
> ``--mca btl ofi,self``).
> 
>   .. important:: Note that if fault tolerance is disabled at runtime,
>                  the ``ft-mpi`` MCA param file is not loaded, thus
>                  components that are unsafe for fault tolerance will
>                  load normally (this may change observed performance
>                  when comparing with and without fault tolerance).

Reviewer: ```suggestion
```

This is wholly redundant with lines 219-232.

* ``mtl``: Matching transport layer Used for MPI point-to-point messages on
  some types of networks

  * All ``mtl`` components are **disabled**

* ``coll``: MPI collective algorithms

  * ``cuda``, ``inter``, ``sync``, ``sm``: **untested** (they have not
    been modified to handle faults, but we expect correct post-fault
    behavior)
  * ``hcoll``, ``portals4`` **disabled** (they have not been modified
    to handle faults, and we expect unspecified post-fault behavior)

* ``osc``: MPI one-sided communications

  * All ``osc`` components are **untested** (they have not been
    modified to handle faults, and we expect unspecified post-fault
    behavior)

* ``io``: MPI I/O and dependent components

  * ``fs``: File system functions for MPI I/O
  * ``fbtl``: File byte transfer layer: abstraction for individual
    read/write operations for OMPIO
  * ``fcoll``: Collective read and write operations for MPI I/O
  * ``sharedfp``: Shared file pointer operations for MPI I/O
  * All components in these frameworks are unmodified, **untested**
    (we expect clean post-failure abort)

* ``vprotocol``: Checkpoint/Restart components

>   * These components have not been modified to handle faults, and are

Reviewer: ```suggestion
  * All ``vprotocol`` components have not been modified to handle faults, and are
```

  reported as failed with the Open MPI failure detector.
  This option is only relevant when ``mpi_ft_detector`` is ``true``.
* ``mpi_ft_detector_period <float> (default: 3e0 seconds)`` heartbeat
  period. Recommended value is 1/3 of the timeout. _Values lower than
  100us may impart a noticeable effect on latency (typically a 3us
  increase)._
  This option is only relevant when ``mpi_ft_detector`` is ``true``.
* ``mpi_ft_detector_timeout <float> (default: 1e1 seconds)`` heartbeat
  timeout (i.e. failure detection speed). Recommended value is 3 times
  the heartbeat period.
  This option is only relevant when ``mpi_ft_detector`` is ``true``.

Known Limitations in ULFM
-------------------------

* InfiniBand support is provided through the UCT BTL; fault tolerant
  operation over the UCX PML is not yet supported for production runs.
* TOPO, FILE, RMA are not fault tolerant. They are expected to work
  properly before the occurence of the first failure.

Modified, Untested and Disabled Components
------------------------------------------

Frameworks and components are listed below and categorized into one of
three classifications:

> 1. **Modified:** This framework/component has been specifically modified
>                  such that it will continue to work after a failure.
> 2. **Untested:** This framework/component has not been modified and/or
>                  tested with fault tolerance scenarios, and _may_ malfunction
>                  after a failure.
> 3. **Disabled:** This framework/component will cause unspecified behavior when
>                  fault tolerance is enabled.

Reviewer: ```suggestion
1. **Modified:** This framework/component has been specifically modified
    such that it will continue to work after a failure.
2. **Untested:** This framework/component has not been modified and/or
   tested with fault tolerance scenarios, and _may_ malfunction
   after a failure.
3. **Disabled:** This framework/component will cause unspecified behavior when
    fault tolerance is enabled.
```

This doesn't render the bullets correctly with the additional spacing on the 2nd+3rd lines after the bullet start.
